### PR TITLE
add leaderElection

### PR DESCRIPTION
**Pull request by M00nF1sh**
Add leaderElection.
The election will be done by an configMap named `app-mesh-controller-leader` by default.
Test done:
1. two replicas running, only one will continue to run, while another block on leadership election
2. turn down the leader pod, the other standby pod becomes leader
3. modify the configmap `app-mesh-controller-leader`, the leader will quit after logging 'losing leader' and restart by kubelet

Helm chart change for added permission:  https://github.com/aws/eks-charts/pull/70/files

Some note: 
       1. configMap is favored over endpoints for leader election. 
       2. lease object is favored over configMap. but it's only available after 1.12.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -198,21 +224,76 @@ func (c *Controller) Run(threadiness int, stopCh chan struct{}) error {
...
+		LeaseDuration: 60 * time.Second,
+		RenewDeadline: 15 * time.Second,
+		RetryPeriod:   5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				c.runWorkers(ctx, threadiness)
+			},
+			OnStoppedLeading: func() {
+				klog.Info("losing leader")
+			},
```
Comment (mogren) [inline pkg/controller/controller.go]: Will the workers stop automatically if it's no longer the leader? Where is `stopCh` closed?

File: pkg/controller/controller_util.go [hunk 2]
```diff
@@ -62,3 +67,22 @@ func namespacedResourceName(resourceName string, defaultResourceNamespace string
 	}
 	return resourceName + "-" + defaultResourceNamespace
 }
+
+// getInClusterNamespace returns the namespace of controller pod.
+func getInClusterNamespace() (string, error) {
+	// Check whether the namespace file exists.
+	// If not, we are not running in cluster so can't guess the namespace.
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply pkg/controller/controller_util.go]: 🤣  changed them, nice recommendations, haha

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -198,21 +224,76 @@ func (c *Controller) Run(threadiness int, stopCh chan struct{}) error {
 		return fmt.Errorf("failed to wait for caches to sync")
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply pkg/controller/controller.go]: 👀

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -198,21 +224,76 @@ func (c *Controller) Run(threadiness int, stopCh chan struct{}) error {
...
+			Identity: leaderElectionIdentity,
+		})
+	if err != nil {
+		return err
+	}
+	elector, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          leaderElectionLock,
+		LeaseDuration: 60 * time.Second,
+		RenewDeadline: 15 * time.Second,
+		RetryPeriod:   5 * time.Second,
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply pkg/controller/controller.go]: yes, i got it from https://github.com/kubernetes/client-go/blob/master/examples/leader-election/main.go 👀

File: pkg/controller/controller.go (hunk 1 above)
Comment (M00nF1sh) [reply pkg/controller/controller.go]: yes, the stopCh didn't close, but the process will finish. e.g. the elector.Run(ctx) will finish :D

**Pull request merged**